tfx/orchestration/airflow/airflow_dag_runner.py

```diff
@@ -111,3 +114,26 @@ def run(self, tfx_pipeline: pipeline.Pipeline):
             component_impl_map[upstream_node])
 
     return airflow_dag
+
+
+  def _replace_runtime_params(self, comp):
+    for k, prop in comp.spec.exec_properties.copy().items():
+      if isinstance(prop, RuntimeParameter):
+        
+        # Airflow only supports string parameters
+        if prop.ptype != str:
+          raise RuntimeError(
+            f'RuntimeParameter in Airflow does not support {prop.ptype} only ptype supported is string'
+          )
+        
+        # If default is a template, we can just remove the template markers to make it work.
+        if prop.default.startswith('{{') and prop.default.endswith('}}'):
+          default = prop.default.replace('{{','').replace('}}', '')
+        else:
+          default = f'"{prop.default}"'
+        
+        # Once we move to support Airflow 2.0 we should instead use airflow.models.DagParam class 
+        comp.spec.exec_properties[k] = f'{{{{ dag_run.conf.get("{prop.name}", {default}) }}}}'
```

Comment:
One general question about the way AirflowDagRunner works right now -- is it the case that if a string execution property is passed with "{{...}}" inside it, the contents inside the braces will be parsed and evaluated by Airflow?

CCing people who can comment on the original intent: @ruoyu90 @1025KB @zhitaoli.  Even if this was not intended, this is the implicit contract now for the AirflowDagRunner, and is a behavior which might not be portable across runners.

Reply:
Assuming the behavior mentioned by @charlesccychen is true: I don't think this behavior can be portable across runner.s. @kennethyang404 is working on an alternative placeholder which can be portable.

I'm okay to leave this decision open, since AirflowDagRunner kinds of predates our current portable efforts and I don't want to over-limit its users if there is a reasonable alternative, but strongly +1 to clearly call out in user doc about this unportable nature.

We can also discuss on possibility of building an alternative of using the placeholders in [1] but I'm not sure if it's worth the effort. WDYT, @casassg ?

[1] https://github.com/tensorflow/tfx/blob/master/tfx/dsl/placeholder/placeholder.py

Reply:
I agree on this not being the best portable option. For example in Kubeflow Pipelines you have an implicit contract when using `[[ScheduledTime]]` which is the equivalent of `{{execution_date}}` in Airflow. This means that I need to change my pipeline when I change orchestrator.

While I think a Placeholder may work, we need to add logic on DagRunners to map it to the underlying orchestrator native representation at some point, given that this is an execution parameter from the orchestrator and not from TFX. Seems currently Placeholder is limited to things under TFX scope.

Still though, I think we should allow users to use the underlying mechanism here. If user wants to use native templating system built into Kubeflow or Airflow, we should not stop them from doing so either. Replicating those options fully from TFX may take some time and even probs be completely impossible to fully replicate as some features may not be available in all orchestrators (example: Airflow has full blown Jinja templating mechanism, Kubeflow only has some limited template options).

I'm happy to discuss this more in depth in person as well :D